### PR TITLE
GKE enhancement of some complex options

### PR DESCRIPTION
**Pull request by fyery-chen**
Problem:
We should expose some extra parameters for options, because some customers need them.

Solved:
GKE enhancement to add more complex options

Issues: 

https://github.com/rancher/rancher/issues/16411
https://github.com/rancher/rancher/issues/10922
https://github.com/rancher/rancher/issues/13980
https://github.com/rancher/rancher/issues/14298
https://github.com/rancher/rancher/issues/14634

UI Issues:

https://github.com/rancher/rancher/issues/16412
https://github.com/rancher/rancher/issues/17990
https://github.com/rancher/rancher/issues/17991
https://github.com/rancher/rancher/issues/17992
https://github.com/rancher/rancher/issues/17993

Related PR:
https://github.com/rancher/kontainer-engine/pull/106

### CODE REVIEW

File: pkg/controllers/management/clusterprovisioner/provisioner.go [hunk 1]
```diff
@@ -244,7 +244,8 @@ func (p *Provisioner) reconcileCluster(cluster *v3.Cluster, create bool) (*v3.Cl
 			logrus.Infof("Create done, Updating cluster [%s]", cluster.Name)
 			apiEndpoint, serviceAccountToken, caCert, err = p.driverUpdate(cluster, *spec)
 		}
-	} else if spec.RancherKubernetesEngineConfig.RotateCertificates != nil {
+	} else if spec.RancherKubernetesEngineConfig != nil &&
+		spec.RancherKubernetesEngineConfig.RotateCertificates != nil {
```
Comment (cjellick) [inline pkg/controllers/management/clusterprovisioner/provisioner.go]: why this change? how is it related to gke options?
if you are fixing a general nil pointer bug, thats fine, but it shoiuld be a separate commit

Comment (fyery-chen) [reply pkg/controllers/management/clusterprovisioner/provisioner.go]: Yeah, I am fixing a nil pointer bug, I will add a commit separately, thanks!

Comment (fyery-chen) [reply pkg/controllers/management/clusterprovisioner/provisioner.go]: This bug has been fixed in the master, so I drop it.

File: server/capabilities/gke_networks_endpoint.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
+package capabilities
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"google.golang.org/api/compute/v1"
```
Comment (nathan-jenan-rancher) [inline server/capabilities/gke_networks_endpoint.go]: Put this import below the golang imports separated by a space, like:
```
        "context"
 	"encoding/json"
 	"fmt"
        "net/http"

 	"google.golang.org/api/compute/v1"
```

File: server/capabilities/gke_service_accounts_endpoint.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
+package capabilities
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"google.golang.org/api/iam/v1"
+
+	"net/http"
```
Comment (nathan-jenan-rancher) [inline server/capabilities/gke_service_accounts_endpoint.go]: Move this import to be next to the 3 other golang imports.

File: server/capabilities/gke_subnetworks_endpoint.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
+package capabilities
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"google.golang.org/api/compute/v1"
```
Comment (nathan-jenan-rancher) [inline server/capabilities/gke_subnetworks_endpoint.go]: Separate this import out from the other golang imports.

File: server/capabilities/gke_networks_endpoint.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+		return
+	}
+
+	client, err := g.getServiceClient(context.Background(), body.Credentials)
+
+	if err != nil {
+		writer.WriteHeader(http.StatusInternalServerError)
+		handleErr(writer, err)
+		return
+	}
```
Comment (nathan-jenan-rancher) [inline server/capabilities/gke_networks_endpoint.go]: I see a duplication of code between all the GKE request up until about this point, can we extract that out into some common code for the various endpoints?

File: server/capabilities/gke_networks_endpoint.go (hunk 2 above)
Comment (fyery-chen) [reply server/capabilities/gke_networks_endpoint.go]: Updated!

File: server/capabilities/gke_service_accounts_endpoint.go (hunk 3 above)
Comment (fyery-chen) [reply server/capabilities/gke_service_accounts_endpoint.go]: Updated!

File: server/capabilities/gke_subnetworks_endpoint.go (hunk 4 above)
Comment (fyery-chen) [reply server/capabilities/gke_subnetworks_endpoint.go]: Updated!

File: server/capabilities/gke_networks_endpoint.go (hunk 5 above)
Comment (fyery-chen) [reply server/capabilities/gke_networks_endpoint.go]: Yep, this question maybe needs a separate issue, so I will not update it in this PR.

Comment (nathan-jenan-rancher) [reply server/capabilities/gke_networks_endpoint.go]: If you don't want to refactor existing funcs I'm fine with that, but you are adding a lot of duplicated code in this PR and I don't want to LGTM a PR like that unless we have a compelling reason to.

Comment (cjellick) [reply server/capabilities/gke_networks_endpoint.go]: should be addressed in this pr

Comment (fyery-chen) [reply server/capabilities/gke_networks_endpoint.go]: OK, I will address it in this pr and as for those existing code duplication, I will add a separate commit.

**Comment by cjellick** [comment]
LGTM

**Comment by cjellick** [comment]
Is there a UI isue for this? I do not see one

**Comment by loganhz** [comment]
Added related UI issues

File: server/capabilities/gke_zones_endpoint.go [hunk 6]
```diff
@@ -17,35 +15,12 @@ type GKEZonesHandler struct {
...
-
-	err = validateCapabilitiesRequestBody(writer, &body.capabilitiesRequestBody)
-
-	if err != nil {
-		handleErr(writer, err)
-		return
-	}
-
-	client, err := g.getServiceClient(context.Background(), body.Credentials)
+	client, err := getComputeServiceClient(context.Background(), body.Credentials)
```
Comment (orangedeng) [inline server/capabilities/gke_zones_endpoint.go]: Use the context from `*http.Request`

**Comment by loganhz** [comment]
Close in favor of https://github.com/rancher/rancher/pull/18006
